### ISSUE TITLE
asyncio support in mux

### ISSUE THREAD

**Issue by bmcfee** [bmcfee/pescador#30]
The mux could benefit from asyncio.  We'd have to be clever about scheduling, but i think it's possible.

**Comment by cjacoby** [comment]
I've done some work with asyncio recently. ...HAMR project?

**Comment by bmcfee** [comment]
If you want to take the lead on it, for sure!

I'm planning to sprint on librosa stuff, but happy to help out here as well.

**Comment by cjacoby** [comment]
Happy do that, if we agree on specs/design/a plan/vision.

**Comment by bmcfee** [comment]
:+1: let's do it.

As for specs... it's a rather complicated beast.  If you've ever looked inside mux, well... I apologize.


Mux is actually doing a few different things:

1. managing live/dead streams
2. collecting samples into a common stream
3. sampling from the common stream (with non-uniform weights)

Point 3 will be the trickiest for concurrent execution, I think.  However, in the current implementation, points 2 and 3 are intertwined.  I think we'll have to factor those apart to make progress here.

Maybe the way to go is to first stage samples from the streams into a common pool, and attach weights according to their originating stream.  We can then sample from the common pool according to the weights, and everything ought to work out to the same distribution, provided that all seeds have equal (expected) latency.  Maybe there's a fancier strategy to be implemented here that maintains the proper sampling distribution regardless of latency?

A few other things to worry about:

- Bounding the size of the common pool.
- Replenishing samples.  I'm not sure asyncio.gather is the right way to do this; i want to pull randomly from any coroutine that's already yielded.  Maybe we just need to use an `asyncio.Queue` for this?
- Any thoughts about the stream management logic?
-

**Comment by cjacoby** [comment]
Okay, I did some homework and spent some time understanding how mux works.

Can you talk a bit about:
1. What the objective of these changes are?
2. How you would interface with them? i.e. Would the API change, or would all of this (including the event loop) be wrapped up inside?

Either way, we (obviously, I think) need to benchmark the current approach versus the new approach to see if it is improved.

I'll do some homework on `asyncio.gather` and `asyncio.Queue` this week.

**Comment by bmcfee** [comment]
> 1. What the objective of these changes are?

The current algorithm works as follows:

1. Activate a number of streams
2. Until done:
    a. select a stream at random (according to the weights)
    b. sample from that stream

The problem here is that step 2b can introduce latency into the process when the stream stalls for some reason.  Stalls happen when the stream needs to access disk, or be restarted/replaced due to a `StopIteration`.

What I'd like is for stream sampling to be decoupled from multiplexing, and the two can happen concurently.  Streams can push into a queue as data becomes available, and the mux will always have access to fresh data.

This gets complicated when we need to deal with weighted sampling; we might need a fancier data structure to handle this properly.  For now though, let's worry about uniform sampling.



> 2. How you would interface with them?

Ideally, the external API would look the same.  However, we might need to implement it as a separate function in order to keep the current implementation running on python<3.5.



> Either way, we (obviously, I think) need to benchmark the current approach versus the new approach to see if it is improved.

Agreed.  I suspect @ejhumphrey has thoughts about that?

**Comment by bmcfee** [comment]
I think this could impact #49.

What do yall @ejhumphrey @cjacoby think about dropping support for legacy python so that we can re-design the mux entirely around asyncio in 1.0?

**Comment by cjacoby** [comment]
I am strongly in favor of dropping support for legacy python.

And excited for that redesign! Have been working on asyncio stuff a bit at work in the past few months, and while it is still somewhat opaque in documentation, I like it.

**Comment by bmcfee** [comment]
Okay.. well, given [pep525](https://docs.python.org/3/whatsnew/3.6.html), it sounds like you can't mix asyncio with generators until python 3.6?  In that case, would this mean that pescador becomes >=3.6 only?  Or is there a clever way to work around this limitation so we can still support 3.4/5?

**Comment by ejhumphrey** [comment]
I have mixed feelings about dropping legacy (<2.7) support. On the one hand, hardlines are the only thing that will really drive adoption, and doing this sooner than later will make it less painful for any users who aren't prepared for the inevitable ... on the other, we're still maybe 1-2 years away from folks taking the EoL seriously, and I worry about who gets left behind in the process (e.g. academics who don't control their runtime environments on compute clusters).

So, my conflicted stance makes me an abstention.

update, seeing @bmcfee's comment: >=3.6 is obvs not gonna fly, and I'm a big fan of cleverness.

**Comment by bmcfee** [comment]
@ejhumphrey alternately, we maintain two implementations of mux, and use a runtime version check to determine which one to import.  I hate the idea of maintaining two copies of the same class, but it's probably the cleanest way to preserve 2.7 support.

**Comment by ejhumphrey** [comment]
well what if we first write a subpackage as a dependency called `byepy2` that throws a deprecation warning for python2.7 any time it's imported?

**Comment by bmcfee** [comment]
> well what if we first write a subpackage as a dependency called byepy2 that throws a deprecation warning for python2.7 any time it's imported?

Seems overkill to have a submodule consisting of `if six.PY2: ...`, but yolo.

Otherwise, I don't mind having the old-style synchronous mux be deprecated, and only used as a fallback.

**Comment by ejhumphrey** [comment]
it was a joke, I meant the subpackage would warn you that python2 itself is deprecated 😉

**Comment by cjacoby** [comment]
I'm pretty strongly in agreement with this:  https://carreau.github.io/posts/planning-an-early-death-for-python-2.html

Which is to say I'd propose "official" 3.5+ support (because asyncio), but we can still make sure the core library builds with Legacy Python, but perhaps the "more awesome features" can be only python3.

`Be the change...`, etc.

**Comment by bmcfee** [comment]
Agreed; still gotta make sure this is all kosher wrt mixing async and generators in 3.4/3.5 though.

**Comment by cjacoby** [comment]
Yeah; I think the 3.4/3.5/3.6 asyncio changers/messiness is actually where
our problems will lie.

**Comment by bmcfee** [comment]
> Yeah; I think the 3.4/3.5/3.6 asyncio changers/messiness is actually where
> our problems will lie.

Been doing some reading about this. I think it's only a problem if we want to use native coroutines (>=3.5, that is, `async/await/yield from` syntax) instead of generator coroutines (3.4).  I'm not sure if we need native coroutines here, but [this page](http://masnun.com/2015/11/13/python-generators-coroutines-native-coroutines-and-async-await.html) covers some of the differences.

I think this might be easiest to implement first using 3.6, and then back-port to workaround async generator syntax.

**Comment by bmcfee** [comment]
(update after scipy2017)

An alternative might be to re-implement mux using [concurrent futures](https://docs.python.org/3/library/concurrent.futures.html).  Even within a single python process, we can still get concurrency over the individual streams using the `ThreadPoolExecutor`, and have the mux sample data from the streams with live data.

This wouldn't make sense for things like round-robin mux, but it should be fine for the more stochastic samplers.  What we'd lose here is the ability to control the exact sampling sequence, since it becomes non-deterministic and dependent on system load, but I think that's generally okay and to be expected.

**Comment by cjacoby** [comment]
That sounds intriguing to be considered/explored, I think, in the general
Mux redesign for 2.0.

**Comment by bmcfee** [comment]
I had a bit of a think on this last night, now that the dust has settled on the refactor.

As noted above, async mux only makes sense for stochastic/shuffled mux, since it could change the order of samples.  A way to make this work is to allow asynchronous streamer execution to push into a buffer, have the mux asynchronously pre-fetch (so that sampling control is still top-down), and then have the mux's iterator just pull from the buffer.  What I'm not sure about is how much (or how deeply) we'd have to modify the existing classes to make this possible.

**Comment by cjacoby** [comment]
@bmcfee I have some cycles to start working ok Pescador again. I'm starting to think about / research / experiment / start to think this; Just curious, before I dive in too deep; do you have any new ideas / thoughts about this you want to share? (Or, do you think there are more important things I should work on?)

My general plan explore / benchmark different strategies for sampling, including:
 • simple asyncio producer / consumer
 • asyncio vs threading vs multiprocessing
 • a "parallel mux", where each "stream" in the mux is it's own thread/process/coroutine, producing via a Queue to the Mux. Possibly with some buffering, to allow there to be samples available in the queue whenever the mux is `.iterate()`d over.

Also planning to write this up as a blog post or notebook or something before I make a definitive pull request solution.

**Comment by bmcfee** [comment]
This all sounds great.  The only other thing I'd prioritize above this is finishing up the article we mostly wrote, and I think is only wanting for some literature review at this point. :grimacing:
